### PR TITLE
CMake integration

### PR DESCRIPTION
**Pull request by jbmouret**
~~# work in progress~~

The objective is to create and install RobotDartConfig.cmake from waf, so that it is easy to call RobotDart from CMake projects (and still compile/install with waf as this is our prefered tool, which will handle options, etc.).

~~# for now~~

~~I managed to use the INTERFACE IMPORTED system to create a working RobotDartConfig.~~

~~# now~~

~~Copy information from dart to this (easy).~~

**Edit:** It seems like a first working version..

### CODE REVIEW

**Comment by costashatz** [comment]
@jbmouret Nice! How can I try to see how it works?

**Comment by jbmouret** [comment]
You can try.

A `./waf install'

then:
`export CMAKE_PREFIX_PATH=/Users/jmouret/lib/cmake/` (where you installed)

and you can try the example.

**Comment by jbmouret** [comment]
Apparently my last commit added many files that should be removed. I will clean up.

**Comment by costashatz** [comment]
> Apparently my last commit added many files that should be removed. I will clean up.

Indeed. It'd be better to revert the commit and only add the files that you want/need. Otherwise, I think it'll pollute the history quite a lot..

> You can try.
> 
> A `./waf install'
> 
> then:
> `export CMAKE_PREFIX_PATH=/Users/jmouret/lib/cmake/` (where you installed)
> 
> and you can try the example.

It works after a small typo fix (`RobotDartConfig.cmake`->`RobotDARTConfig.cmake`) and only the `no-graphics` version.

**Comment by jbmouret** [comment]
I polluted the history (sorry) because it was already pushed... If you know how to revert the commit and put the files, I do not remember how to do this.

For the graphics version, I have this issue:
```
-- Found DART: /usr/local/include  found components: dart utils utils-urdf collision-bullet collision-ode dart 
CMake Warning at /usr/local/share/cmake/Magnum/FindMagnum.cmake:646 (find_package):
  By not providing "FindGLFW.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "GLFW", but
  CMake did not find one.

  Could not find a package configuration file provided by "GLFW" with any of
  the following names:

    GLFWConfig.cmake
    glfw-config.cmake

  Add the installation prefix of "GLFW" to CMAKE_PREFIX_PATH or set
  "GLFW_DIR" to a directory containing one of the above files.  If "GLFW"
  provides a separate development package or SDK, be sure it has been
  installed.
Call Stack (most recent call first):
  /usr/local/share/cmake/Magnum/MagnumConfig.cmake:26 (include)
  /usr/local/Cellar/cmake/3.17.2/share/cmake/Modules/CMakeFindDependencyMacro.cmake:47 (find_package)
  /Users/jmouret/lib/cmake/RobotDART/RobotDARTConfig.cmake:32 (find_dependency)
  CMakeLists.txt:8 (find_package)


-- Found Magnum: /usr/local/include  found components: Trade GL MeshTools Primitives Shaders SceneGraph GlfwApplication WindowlessCglApplication 
-- Found Magnum: /usr/local/include  found components: Trade GL SceneGraph Primitives MeshTools 
-- Found DART: /usr/local/include (Required is at least version "6.0.0") found components: dart 
-- Found Magnum: /usr/local/include  found components: Trade AnyImageImporter 
-- Configuring done
CMake Error at CMakeLists.txt:17 (add_executable):
  Target "robot_dart_example_graphics" links to target "GLFW::GLFW" but the
  target was not found.  Perhaps a find_package() call is missing for an
  IMPORTED target, or an ALIAS target is missing?


-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
```
@costashatz do you have the same?

**Comment by jbmouret** [comment]
I guess this might be related to: https://github.com/mosra/magnum/pull/194/files

**Comment by costashatz** [comment]
> I polluted the history (sorry) because it was already pushed... If you know how to revert the commit and put the files, I do not remember how to do this.

I remember how, but now it'll be much more complicated. The pollution is not that big in the end :)

> @costashatz do you have the same?

Yes.. This is a known issue with GLFW (they have a very stupid CMake and they do not want to change it). I remember that Magnum did something for this. Let me try to find the solution..

**Comment by costashatz** [comment]
@jbmouret I fixed the issue with GLFW configuration.. Now I am getting linking errors with Magnum. Which makes a bit sense since: 1) we are mixing static and shared libraries together (I have to check how to say to CMake that they are static), and 2) I am not sure Magnum is exporting the targets as you are using them.

**Comment by costashatz** [comment]
@jbmouret it should be fixed now. There were 2 issues: 1) the order of the libraries for linking matters (at least in gcc), and 2) you were not getting the proper variable for Magnum's libraries from waf (thus, they were a few libraries missing). It should be working now..

**Comment by costashatz** [comment]
In my Ubuntu setup, I am getting the following:

```bash
CMake Error at /usr/share/cmake-3.10/Modules/CMakeFindDependencyMacro.cmake:48 (find_package):
  Could not find a package configuration file provided by "Boost" with any of
  the following names:

    BoostConfig.cmake
    boost-config.cmake

  Add the installation prefix of "Boost" to CMAKE_PREFIX_PATH or set
  "Boost_DIR" to a directory containing one of the above files.  If "Boost"
  provides a separate development package or SDK, be sure it has been
  installed
```

**Comment by jbmouret** [comment]
It works for me on my mac! (it compiles and run, except that there my example does not have access to the urdf). 

I am quite sure that we are using targets in the correct ways. Do you have the BoostConfig.cmake somewhere?

```
/usr/local/Cellar/boost/1.72.0_1/lib/cmake/Boost-1.72.0/BoostConfig.cmake
/usr/local/Cellar/boost/1.72.0_1/lib/cmake/Boost-1.72.0/BoostConfigVersion.cmake
```

from brew.

Do you use the system boost or a special one?

**Comment by costashatz** [comment]
> Do you use the system boost or a special one?

The system boost. But the weird thing is the following:

- There's no `BoostConfig.cmake` anywhere installed
- `FindBoost.cmake` is available here: `/usr/local/share/dart/cmake/DARTFindBoost.cmake` and `/usr/share/cmake-3.10/Modules/FindBoost.cmake`

I am not sure if I have installed `dev-all` from Boost, but do we need this?

In my arch-linux system, everything was working properly for Boost.

**Comment by jbmouret** [comment]
According to the cmake documentation:"If Boost was built using the boost-cmake project or from Boost 1.70.0 on it provides a package configuration file for use with find_package’s config mode. "
https://cmake.org/cmake/help/v3.15/module/FindBoost.html 

As this is for all boost (and not a specific component), this might be in some 'dev-all' package, but this would be surprising. This file (BoostConfig.cmake) is in no Ubuntu package?

**Comment by costashatz** [comment]
> This file (BoostConfig.cmake) is in no Ubuntu package?

I do not think that I installed `dev-all`. I think I only installed a few packages. Let me check when I find time.

> from Boost 1.70.0

In Ubuntu 18.04, we have Boost 1.65.1 (in arch-linux, I have 1.72.0).. This might explain the issue. We might need to provide a configuration file if version of Boost is less than 1.70.0..

By the way, I think we should enable components for RobotDART. One for the `Simu` and one for `GUI/Magnum`. I do not like this CMake variable: maybe robot_dart is built with Magnum, but I do not want it. Why would CMake try to find it?

**Comment by jbmouret** [comment]
Yes, this is certainly because you are using an old boost... 

I will have a look later at the components. It should not be hard.

**Comment by costashatz** [comment]
> Yes, this is certainly because you are using an old boost...

This is Ubuntu!

**Comment by costashatz** [comment]
> (it compiles and run, except that there my example does not have access to the urdf).

This is because I am assuming that there files are located next to the executable. We can fix that.

**Comment by costashatz** [comment]
@jbmouret I fixed the issue with Boost easily as I realized that we only use Boost for unit tests (no reason for the user to use them in a client app) and in DART. DART finds on each own Boost, and thus I completely removed it and it works nicely both in archlinux and Ubuntu.

I, then, found another bug in Ubuntu and older versions of CMake that they were ignoring `INTERFACE_LINK_DIRECTORIES`, so I fixed that also.

Finally, I cleaned-up a few things and passed the flags from waf configuration to CMake.

Let me know if the latest commit works for you also. Btw, I like now the components! :smile:

**Comment by jbmouret** [comment]
It works for me! We should select an example that works (fixing the paths of the urdf... or not using any urdf).

I am glad you like the component.

File: cmake/RobotDARTConfigVersion.cmake.in [hunk 1]
```diff
@@ -0,0 +1,11 @@
+set(PACKAGE_VERSION "@robot_dart_VERSION@")
```
Comment (costashatz) [inline cmake/RobotDARTConfigVersion.cmake.in:1]: Is this file used in any way? I guess we should use it to tell to CMake which version of RobotDART we want. I think at the moment it's not being used.. @jbmouret how were you planning on using it?

**Comment by costashatz** [comment]
@jbmouret I tested on my setups and it is working like a charm.. I will merge later today!

**Pull request merged**